backend/src/main/java/ai/verta/modeldb/lineage/LineageDAORdbImpl.java

```diff
@@ -69,166 +293,304 @@ private void validate(List<LineageEntry> inputList, List<LineageEntry> outputLis
+    if (connectionEntities.size() < 2) {
+      switch (connectionEntity.getEntityType()) {
+        case ENTITY_TYPE_EXPERIMENT_RUN:
+          session.delete(
+              session.get(LineageExperimentRunEntity.class, connectionEntity.getEntityId()));
+          break;
+        case ENTITY_TYPE_VERSIONING_BLOB:
+          session.delete(
+              session.get(LineageVersioningBlobEntity.class, connectionEntity.getEntityId()));
+          break;
+      }
+    }
+    session.delete(connectionEntity);
   }
 
-  private void saveOrUpdate(Session session, LineageEntry input, LineageEntry output) {
-    session.saveOrUpdate(new LineageEntity(input, output));
+  private void addLineage(
+      Session session,
+      LineageEntry lineageEntry,
+      Long id,
+      int connectionType,
+      BlobHashInCommitFunction blobHashInCommitFunction)
+      throws ModelDBException {
+    final Long entityId;
+    final int entityType;
+    final boolean connectionExists;
+
+    switch (lineageEntry.getDescriptionCase()) {
+      case EXPERIMENT_RUN:
+        LineageExperimentRunEntity experimentRun =
+            getLineageExperimentRunEntity(session, lineageEntry);
+        if (experimentRun == null) {
+          experimentRun = new LineageExperimentRunEntity(lineageEntry.getExperimentRun());
+          session.save(experimentRun);
+          connectionExists = false;
+        } else {
+          connectionExists = true;
+        }
+        entityId = experimentRun.getId();
+        entityType = ENTITY_TYPE_EXPERIMENT_RUN;
+        break;
+      case BLOB:
+        VersioningLineageEntry blob = lineageEntry.getBlob();
+        InternalFolderElementEntity result = blobHashInCommitFunction.apply(session, blob);
+        LineageVersioningBlobEntity lineageVersioningBlobEntity =
+            getLineageVersioningBlobEntity(
+                session, blob.getRepositoryId(), result.getElement_sha(), result.getElement_type());
+        if (lineageVersioningBlobEntity == null) {
+          lineageVersioningBlobEntity =
+              new LineageVersioningBlobEntity(
+                  blob.getRepositoryId(), result.getElement_sha(), result.getElement_type());
+          session.save(lineageVersioningBlobEntity);
+          connectionExists = false;
+        } else {
+          connectionExists = true;
+        }
+        entityId = lineageVersioningBlobEntity.getId();
+        entityType = ENTITY_TYPE_VERSIONING_BLOB;
+        break;
+      default:
+        throw new ModelDBException("Unknown lineage type");
+    }
+
+    List connectionEntities =
+        getConnectionEntities(session, invert(connectionType), entityId, entityType);
+    if (connectionExists) {
+      if (connectionType == CONNECTION_TYPE_OUTPUT
+          && !((ConnectionEntity) connectionEntities.get(0)).getId().equals(id)) {
+        throw new ModelDBException(
+            "Specified lineage entry already has an output connection", Code.INVALID_ARGUMENT);
+      }
+    }
```

Comment:
What is this checking? Seems wrong, but I had a hard time following this function.

Reply:
It checks if we already have an output to this entry